zkevm-circuits/src/bytecode_circuit/bytecode_chiquito.rs

```diff
@@ -0,0 +1,215 @@
+use std::cell::RefCell;
+
+use bus_mapping::state_db::EMPTY_CODE_HASH_LE;
+use chiquito::{
+    ast::{ToExpr, ToField},
+    compiler::{
+        cell_manager::SingleRowCellManager, step_selector::TwoStepsSelectorBuilder, Compiler,
+    },
+    dsl::circuit,
+    ir::Circuit,
+    stdlib::IsZero,
+};
+use eth_types::Field;
+use halo2_proofs::plonk::{Column, Fixed};
+
+use crate::bytecode_circuit::bytecode_unroller::unroll;
+
+use super::{
+    circuit::{BytecodeCircuitConfigArgs, WitnessInput},
+    wit_gen::BytecodeWitnessGen,
+};
+
+pub fn bytecode_circuit<F: Field>(
+    BytecodeCircuitConfigArgs {
+        bytecode_table,
+        keccak_table,
+        challenges,
+    }: &BytecodeCircuitConfigArgs<F>,
+    push_data_table_value: Column<Fixed>,
+    push_data_table_size: Column<Fixed>,
+) -> Circuit<F, WitnessInput<F>, RefCell<BytecodeWitnessGen<F>>> {
+    let bytecode_circuit = circuit::<F, WitnessInput<F>, RefCell<BytecodeWitnessGen<F>>, _>(
+        "bytecode circuit",
+        |ctx| {
+            use chiquito::dsl::cb::*;
+
+            let length = ctx.forward("length");
+            let push_data_left = ctx.forward("push_data_left");
+            let value_rlc = ctx.forward_with_phase("value_rlc", 1); // 1 -> SecondPhase
+
+            let index = ctx.import_halo2_advice("index", bytecode_table.index);
+            let hash = ctx.import_halo2_advice("hash", bytecode_table.code_hash);
+            let is_code = ctx.import_halo2_advice("is_code", bytecode_table.is_code);
+            let value = ctx.import_halo2_advice("value", bytecode_table.value);
+            let _tag = ctx.import_halo2_advice("tag", bytecode_table.tag); // imported for step selector
+
+            let push_data_table_value =
+                ctx.import_halo2_fixed("push_data_value", push_data_table_value);
+            let push_data_table_size =
+                ctx.import_halo2_fixed("push_data_size", push_data_table_size);
+
+            let keccak_is_enabled =
+                ctx.import_halo2_advice("keccak_is_enabled", keccak_table.is_enabled);
+            let keccak_value_rlc =
+                ctx.import_halo2_advice("keccak_value_rlc", keccak_table.input_rlc);
+            let keccak_length = ctx.import_halo2_advice("keccak_length", keccak_table.input_len);
+            let keccak_hash = ctx.import_halo2_advice("keccak_hash", keccak_table.output_rlc);
+
+            let header = ctx.step_type("header");
+            let byte_step = ctx.step_type("byte");
+
+            ctx.pragma_first_step(header);
+            ctx.pragma_last_step(header);
+
+            ctx.step_type_def(header, move |ctx| {
+                ctx.constr(isz(index));
+                ctx.constr(eq(value, length));
```

Comment:
So the idea, that comes from Edu, is that there are two types of constraints: internal constraints that do not use anything outside the current step instance and transtion constraint that use forward_signal.next(). I think this distinction is the most important.

And then using those two type of constraints with two different methods, use constraint builder functions to express the constraint.

Perhaps we could brainstorm what are the best names for this methods and functions, but I wouldn't just use some names because it is what we used in the past in zkEVM.

Also I could add these functions to a "zkevm-compatibility" trait as syntatic sugar if we consider it is important to keep usint the same names.